File: tests/runners/typescript/tests/test_11_relations_indexes.ts
// Test Case 11: Relations and composite indexes
// Tests reverse relation aliases, FK navigation helpers, and composite indexes.

import { ExamplesRelations } from '../generated/11_relations_indexes/typescript/schema';
import { SchemaContainer } from '../generated/11_relations_indexes/typescript/schema_container';
import {
    SchemaBinaryRefContext,
    type SchemaContainer as BinaryRefSchemaContainer,
} from '../generated/11_relations_indexes/typescript/schema_binary_refs';

console.log("=== Test Case 11: Relations and Composite Indexes ===");

const container = new SchemaContainer({
    Users: [
        {
            id: 1,
            email: "author@example.com",
            displayName: "Author",
        },
    ],
    Posts: [
        {
            id: 10,
            authorId: 1,
            status: ExamplesRelations.PostStatus.Published,
            title: "Relations",
        },
    ],
});

const post = container.Posts.getById(10);
console.assert(post !== undefined, "post should exist");
if (post === undefined) throw new Error("post should exist");

const composite = container.Posts.findByAuthorIdStatus([
    1,
    ExamplesRelations.PostStatus.Published,
]);
console.assert(composite.length === 1, "composite index should match one post");
console.assert(composite[0].id === 10, "composite index should return the post");

const author = container.getPostAuthor(post);
console.assert(author !== undefined, "post author should resolve");
if (author === undefined) throw new Error("post author should resolve");
console.assert(author.email === "author@example.com", "post author email should match");

const posts = container.findUserPosts(author);
console.assert(posts.length === 1, "reverse relation should return one post");
console.assert(posts[0].title === "Relations", "reverse relation should return the post");

const valid = container.validateAll();
console.assert(valid.isValid, "valid relation container should pass FK validation");

const invalid = new SchemaContainer({
    Posts: [
        {
            id: 11,
            authorId: 404,
            status: ExamplesRelations.PostStatus.Draft,
            title: "Missing author",
        },
    ],
});
const invalidResult = invalid.validateAll();
console.assert(!invalidResult.isValid, "missing author should fail FK validation");
console.assert(invalidResult.errorCount === 1, "missing author should produce one FK error");

const binaryContainer: BinaryRefSchemaContainer = {
    Users: [
        {
            id: 1,
            email: "author@example.com",
            displayName: "Author",
        },
    ],
    Posts: [
        {
            id: 10,
            authorId: 1,
            status: ExamplesRelations.PostStatus.Published,
            title: "Relations",
        },
        {
            id: 11,
            authorId: 1,
            status: ExamplesRelations.PostStatus.Draft,
            title: "Draft",
        },
    ],
};

const binaryBytes = SchemaBinaryRefContext.saveBinary(binaryContainer);
const binaryContext = SchemaBinaryRefContext.openBinary(binaryBytes);
const binaryComposite = binaryContext.Posts.findByAuthorIdStatus([
    1,
    ExamplesRelations.PostStatus.Published,
]);
console.assert(binaryComposite.length === 1, "BinaryRef composite index should match one post");
console.assert(binaryComposite[0].title === "Relations", "BinaryRef composite index should return the published post");
console.assert(
    binaryContext.Posts.findByAuthorIdStatus([2, ExamplesRelations.PostStatus.Published]).length === 0,
    "BinaryRef composite index should return empty for a missing tuple",
);

console.log("=== All tests passed! ===");

File: tests/runners/csharp/tests/Test_11_relations_indexes.cs
// Test Case 11: Relations and composite indexes

using System;

class Program
{
    static int passed = 0;
    static int failed = 0;

    static void Assert(bool condition, string message)
    {
        if (!condition)
        {
            Console.WriteLine($"    FAILED: {message}");
            failed++;
        }
    }

    static void TestCompositeIndexAndNavigation()
    {
        Console.WriteLine("  Testing composite index and navigation...");

        var container = new Schema.Container.SchemaDataContainer();
        var user = new examples.relations.User
        {
            id = 1,
            email = "alice@example.com",
            display_name = "Alice"
        };
        var draft = new examples.relations.Post
        {
            id = 10,
            author_id = 1,
            status = examples.relations.PostStatus.Draft,
            title = "Draft post"
        };
        var published = new examples.relations.Post
        {
            id = 11,
            author_id = 1,
            status = examples.relations.PostStatus.Published,
            title = "Published post"
        };

        container.Users.Add(user);
        container.Posts.Add(draft);
        container.Posts.Add(published);

        Assert(container.Posts.ByAuthorId[1].Count == 2, "author group index should include both posts");
        var publishedByComposite = container.Posts.ByAuthorIdStatus[(1u, examples.relations.PostStatus.Published)];
        Assert(publishedByComposite.Count == 1, "composite index should filter by author and status");
        Assert(publishedByComposite[0].id == 11, "composite index should return published post");
        Assert(user.Posts.Count == 2, "reverse navigation should resolve user posts");
        Assert(published.author?.email == "alice@example.com", "forward FK navigation should resolve author");
        container.ValidateOrThrow();

        var path = System.IO.Path.Combine(
            System.IO.Path.GetTempPath(),
            "polygen-csharp-relations-binary-ref-" + Guid.NewGuid().ToString("N") + ".bin");
        try
        {
            Schema.BinaryRefs.SchemaBinaryRefContext.SaveBinary(path, container);
            var ctx = Schema.BinaryRefs.SchemaBinaryRefContext.OpenBinary(path);

            var binaryPublished = ctx.Posts.FindByAuthorIdStatus((1u, examples.relations.PostStatus.Published));
            Assert(binaryPublished.Count == 1, "BinaryRef composite index should filter by author and status");
            Assert(binaryPublished[0].id == 11, "BinaryRef composite index should return published post");
            Assert(ctx.Posts.FindByAuthorIdStatus((1u, examples.relations.PostStatus.Archived)).Count == 0,
                "BinaryRef composite index should return empty list for missing tuple");
        }
        finally
        {
            if (System.IO.File.Exists(path))
            {
                System.IO.File.Delete(path);
            }
        }

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestMissingForeignKeyValidation()
    {
        Console.WriteLine("  Testing missing FK validation...");

        var container = new Schema.Container.SchemaDataContainer();
        container.Posts.Add(new examples.relations.Post
        {
            id = 20,
            author_id = 999,
            status = examples.relations.PostStatus.Archived,
            title = "Broken post"
        });

        var result = container.ValidateAll();
        Assert(!result.IsValid, "missing author should fail validation");
        Assert(result.Errors.Count == 1, "missing author should report one validation error");
        Assert(result.Errors[0].ConstraintType == "ForeignKey", "validation error should be ForeignKey");

        var threw = false;
        try
        {
            container.ValidateOrThrow();
        }
        catch (Polygen.Common.ValidationException)
        {
            threw = true;
        }
        Assert(threw, "ValidateOrThrow should reject missing FK");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void Main()
    {
        Console.WriteLine("=== Test Case 11: Relations and composite indexes ===");

        TestCompositeIndexAndNavigation();
        TestMissingForeignKeyValidation();

        if (failed > 0)
        {
            Console.WriteLine($"=== {failed} tests failed! ===");
            Environment.Exit(1);
        }

        Console.WriteLine("=== All tests passed! ===");
    }
}
